### PR TITLE
Fix issues

### PR DESCRIPTION
**Pull request by fryerd1**
## Description
Fix formatting and wording errors, add INSTRUCTIONS.md, remove bonus unit test exercises, add reports folder and change report_path to report_dir

## Type of change

*You can delete options that are not relevant.*

- [ ] Bug fix - *non-breaking change*
- [ ] New feature - *non-breaking change*
- [ ] Breaking change - *backwards incompatible change, changes expected behaviour*
- [ ] Non-user facing change, structural change, dev functionality, docs ...

## Checklist:

- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code appropriately, focusing on explaining my design decisions (explain why, not how).
- [ ] I have made corresponding changes to the documentation (comments, docstring, etc.. )
- [ ] I have added tests that prove my fix is effective or that my feature works (see here for more information).
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have updated the changelog.
- [ ] I have checked the pipeline runs with test data.

<br>

#  Peer review
Any new code includes all the following:

- **Documentation**: docstrings, comments have been added/ updated.
- **Style guidelines**: New code conforms to the project's contribution guidelines.
- **Functionality**: The code works fully implements the requirements and works as expected, handles expected edge cases, exceptions are handled appropriately.
- **Complexity**: The code is not overly complex, logic has been split into appropriately sized functions, etc..
- **Test coverage**: Unit tests cover essential functions for a reasonable range of inputs and conditions. Added and existing tests pass on my machine.

### Review comments
Suggestions should be tailored to the code that you are reviewing. Provide context.
Be critical and clear, but not mean. Ask questions and set actions.
<details><summary>These might include:</summary>

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented
  - Do the tests effectively assure that it
  works correctly? Are there additional edge cases/ negative tests to be considered?
- code style improvements (could the code be written more clearly?)
</details>
<br>

*Further reading: [code review best practices](https://best-practice-and-impact.github.io/qa-of-code-guidance/peer_review.html)*

### CODE REVIEW

File: exercises/solutions/02_modules_solutions.ipynb [hunk 1]
```diff
@@ -30,8 +30,8 @@
     "from python_rap_demo.utils import add_bmi_column\n",
     "\n",
     "input_path = \"../../data/input/health_data.csv\"\n",
-    "cleaned_path = \"../outputs/02_modules/cleaned_data.csv\"\n",
-    "report_path = \"../outputs/02_modules/\"\n",
+    "cleaned_path = \"../data/outputs/cleaned/health_data_cleaned.csv\"\n",
+    "report_path = \"../reports/disease_prevalence_report.md\"\n",
```
Comment (alex-westwood) [inline exercises/solutions/02_modules_solutions.ipynb]: I'm getting some errors running the chart solution. Believe just changing this should fix it


```suggestion
    "report_path = \"../../reports/disease_prevalence_report.md\"\n",
```

File: exercises/solutions/04_unit_tests_solutions.ipynb [hunk 2]
```diff
@@ -15,7 +15,7 @@
    "id": "1",
    "metadata": {},
    "source": [
-    "## Solution 1: Write a simple unit test for a new function\n",
+    "## Exercise 2 Solution: Write a simple unit test for a new function\n",
```
Comment (alex-westwood) [inline exercises/solutions/04_unit_tests_solutions.ipynb:102]: Thanks for adding exercise 1, I like it. I think its worth adding a solution for it here even if it is just exploratory, as some people may be completely new to unit tests. The solution could just reiterate where to find the relevant function and unit test, give instructions on how to run the test, and then give an example of how to modify the test df and what error that creates

File: exercises/04_unit_tests.ipynb [hunk 3]
```diff
@@ -31,7 +31,22 @@
...
    "source": [
-    "## Exercise 1: Write a simple unit test for a new function\n",
+    "## Exercise 1: Review an existing unit test\n",
+    "\n",
+    "The function `clean_health_data` in `src/python_rap_demo/cleaning.py` has an existing unit test in\n",
+    "`tests/test_cleaning.py`.\n",
+    "\n",
+    "**Task:** \n",
+    "1. Run the existing unit test for `clean_health_data` to understand how it works. \n",
+    "2. Modify the sample DataFrame in the test to understand what causes the tests to pass or fail (Note: if all rows are dropped all tests will fail)."
```
Comment (alex-westwood) [inline exercises/04_unit_tests.ipynb]: This is great thank you!

File: INSTRUCTIONS.md [hunk 4]
```diff
@@ -0,0 +1,8 @@
+Offline Python RAP demo instructions
+
+Getting started
+
+1. Download and unzip the file
+2. Open the folder in your chosen IDE
+
+Once these steps are complete refer to README.md step 2 and onwards
```
Comment (alex-westwood) [inline INSTRUCTIONS.md]: Thanks for this Dylan. One small suggestion is to reword this to say where the README.md is e.g. 


```suggestion
Once these steps are complete refer to the README.md file in the unzipped folder and follow the instructions from step 2 onwards
```

Just in case people go looking for a README on the platform somewhere.

Agree these instructions don't need to be on GitHub as well, we'll add it to the tool page on the platform

**Review by alex-westwood — CHANGES_REQUESTED** [review]
Thanks for this. It looks great. I've added a couple suggestions, but let me know if any q's on them

File: exercises/solutions/04_unit_tests_solutions.ipynb [hunk 5]
```diff
@@ -15,7 +15,7 @@
...
     "\n",
     "Here are example unit tests for the `flag_missing` and `impute_by_group` functions in `src/python_rap_demo/cleaning.py`:"
```
Comment (alex-westwood) [inline exercises/solutions/04_unit_tests_solutions.ipynb:104]: The code blocks for flag_missing and impute_by_group tests in the solutions won't run unless they are copied into cleaning.py. Can you add a note to that effect.

Also can you test out copying and pasting the functions into cleaning.py to see if those solution code blocks run. I think we may also need to add 

```python
import os
import sys

sys.path.insert(0, os.path.abspath(os.path.join(os.getcwd(), "..", "..", "src")))
```

in a setup block similar to the modules solutions

File: INSTRUCTIONS.md [hunk 6]
```diff
@@ -0,0 +1,8 @@
+Offline Python RAP demo instructions
+
+Getting started
+
+1. Download and unzip the file
+2. Open the folder in your chosen IDE
+
+Once these steps are complete, refer to the README.md file in the unzipped folder and follow the instrcutions from step 2 onwards
```
Comment (alex-westwood) [inline INSTRUCTIONS.md]: typo

```suggestion
Once these steps are complete, refer to the README.md file in the unzipped folder and follow the instructions from step 2 onwards
```

File: exercises/solutions/04_unit_tests_solutions.ipynb (hunk 5 above)
Comment (alex-westwood) [reply exercises/solutions/04_unit_tests_solutions.ipynb:104]: yep perfect thank you!

File: exercises/solutions/04_unit_tests_solutions.ipynb (hunk 2 above)
Comment (alex-westwood) [reply exercises/solutions/04_unit_tests_solutions.ipynb:102]: I was thinking slightly different for this one. Will call to chat about it

**Review by alex-westwood — CHANGES_REQUESTED** [review]
Worth having a chat on the unit test one, as I don't think I was clear, but once thats done it will be ready to merge in

File: exercises/solutions/04_unit_tests_solutions.ipynb [hunk 7]
```diff
@@ -15,9 +15,13 @@
...
    "source": [
-    "## Solution 1: Write a simple unit test for a new function\n",
+    "## Exercise 1 Solution: Review and Adapt an existing unit\n",
     "\n",
-    "Here are example unit tests for the `flag_missing` and `impute_by_group` functions in `src/python_rap_demo/cleaning.py`:"
+    "Open `tests/test_cleaning.py` and run the test using the following command in the terminal:\n",
+    "```cmd\n",
+    "pytest tests/test_cleaning.py\n",
+    "```\n",
+    "After the test passes, change the DataFrame to the following DataFrame and run the test again:"
```
Comment (alex-westwood) [inline exercises/solutions/04_unit_tests_solutions.ipynb]: Suggest adding where to change the DataFrame here e.g.


```suggestion
    "After the test passes, change the DataFrame in `test_cleaning.py` to the following DataFrame and run the test again:"
```

File: exercises/solutions/04_unit_tests_solutions.ipynb [hunk 8]
```diff
@@ -26,10 +30,125 @@
...
+    "    }\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3",
+   "metadata": {},
+   "source": [
+    "Notice how the check for the missing smoker value fails. The test checks column 0 for a \"smoker\" value of \"No\", with the new DataFrame the first 2 columns are imputed due to missing diagnosis leaving the DataFrame looking like this:"
```
Comment (alex-westwood) [inline exercises/solutions/04_unit_tests_solutions.ipynb]: I like how this is set out, although seeing it now, wondering if it would be clearer to change the fillna in the cleaning.py function from "No" to "Yes" as thats more likely to be why a unit test fails in practice. 

Can then explain that the unit test fails because "Yes" is being imputed into the None values in the test dataframe and that the unit test failing highlights a key change in functionality of the pipeline for the developer to check.

Not sure if thats clear so happy to have a quick chat about this tomorrow

Comment (alex-westwood) [reply exercises/solutions/04_unit_tests_solutions.ipynb]: If not, I think your suggestion of the first column works better than column 0. And use the full word for numbers e.g.

```suggestion
    "Notice how the check for the missing smoker value fails. The test checks the first column for a \"smoker\" value of \"No\", with the new DataFrame the first two columns are imputed due to missing diagnosis leaving the DataFrame looking like this:"
```

**Review by alex-westwood — CHANGES_REQUESTED** [review]

File: exercises/solutions/04_unit_tests_solutions.ipynb [hunk 9]
```diff
@@ -15,9 +15,13 @@
    "id": "1",
    "metadata": {},
    "source": [
-    "## Solution 1: Write a simple unit test for a new function\n",
+    "## Exercise 1 Solution: Review and Adapt an existing unit\n",
```
Comment (alex-westwood) [inline exercises/solutions/04_unit_tests_solutions.ipynb]: typo
```suggestion
    "## Exercise 1 Solution: Review and adapt an existing unit test\n",
```

File: exercises/solutions/04_unit_tests_solutions.ipynb [hunk 10]
```diff
@@ -26,10 +30,95 @@
...
+    "\n",
+    "    return df"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3",
+   "metadata": {},
+   "source": [
+    "Notice how the check for the missing smoker value fails. The test checks the first column for a \"smoker\" value of \"No\", however the modified `clean_health_data` function fills missing smoker values with 'Yes', changing the smoker value in the first column to 'Yes', which causes the test to fail.\n",
```
Comment (alex-westwood) [inline exercises/solutions/04_unit_tests_solutions.ipynb:63]: Love this

File: exercises/04_unit_tests.ipynb [hunk 11]
```diff
@@ -31,7 +31,55 @@
    "id": "2",
    "metadata": {},
    "source": [
-    "## Exercise 1: Write a simple unit test for a new function\n",
+    "## Exercise 1: Review and Adapt an existing unit test\n",
```
Comment (alex-westwood) [inline exercises/04_unit_tests.ipynb:34]: Just to remove the capital A

```suggestion
    "## Exercise 1: Review and adapt an existing unit test\n",
```

File: exercises/04_unit_tests.ipynb [hunk 12]
```diff
@@ -31,7 +31,55 @@
...
+    "\n",
+    "The function `clean_health_data` in `src/python_rap_demo/cleaning.py` has an existing unit test in\n",
+    "`tests/test_cleaning.py`.\n",
+    "\n",
+    "**Task:** \n",
+    "1. Run the existing unit test for `clean_health_data` to understand how it works. \n",
+    "2. Modify the function `clean_health_data` and re-run the tests to understand what causes the tests to pass or fail, then modify the unit test so that the tests passes again.\n",
+    "\n",
+    "There is an example of a modified function below however it does not have to be used, feel free to modify the function and experiment with the tests without following the set exercise."
+   ]
```
Comment (alex-westwood) [inline exercises/04_unit_tests.ipynb:44]: Brilliant, thank you :)

File: exercises/solutions/04_unit_tests_solutions.ipynb [hunk 13]
```diff
@@ -26,10 +30,95 @@
...
+    "\n",
+    "In order for the test to pass change the expected \"smoker\" value to \"Yes\" instead of \"No\".\n",
```
Comment (alex-westwood) [inline exercises/solutions/04_unit_tests_solutions.ipynb]: Suggest adding a bit more detail here e.g. 

```suggestion
    "The test failing highlights the function to developers who can then check if the change was correct or not. If it was not, the developer can fix the error in the function. If it was, the unit test can be adapted to incorporate the change. In this case assume the change was correct. In order for the test to pass change the expected \"smoker\" value to \"Yes\" instead of \"No\".\n",
```

**Review by alex-westwood — APPROVED** [review]
